CMakeLists.txt

```diff
@@ -158,8 +164,80 @@ if(OCOS_ENABLE_SELECTED_OPLIST)
   include(_selectedoplist)
 endif()
 
+set(_OCOS_EXCEPTIONS_REQUIRED OFF)
+if (OCOS_ENABLE_GPT2_TOKENIZER OR
+    OCOS_ENABLE_WORDPIECE_TOKENIZER OR
+    OCOS_ENABLE_BLINGFIRE OR
+    OCOS_ENABLE_SPM_TOKENIZER OR
+    (OCOS_ENABLE_CV2 OR OCOS_ENABLE_OPENCV_CODECS OR OCOS_ENABLE_VISION))
+    set(_OCOS_EXCEPTIONS_REQUIRED ON)
+endif()
+
+# Special case an embedded build with ORT exceptions disabled but custom ops that require exceptions.
+# Allow using an override so we can do a direct build of ort-ext in a CI without having to embed it in an ORT build.
+set(_OCOS_PREVENT_EXCEPTION_PROPAGATION OFF)
+if (_OCOS_PREVENT_EXCEPTION_PROPAGATION_OVERRIDE)
+  set(_OCOS_PREVENT_EXCEPTION_PROPAGATION ${_OCOS_PREVENT_EXCEPTION_PROPAGATION_OVERRIDE})
+elseif(_ONNXRUNTIME_EMBEDDED AND onnxruntime_DISABLE_EXCEPTIONS AND _OCOS_EXCEPTIONS_REQUIRED)
+  set(_OCOS_PREVENT_EXCEPTION_PROPAGATION ON)
+endif()
+
+if (_OCOS_PREVENT_EXCEPTION_PROPAGATION)
+  message(STATUS "Embedded build as part of ONNX Runtime with exceptions disabled. "
+                 "Extensions will be built with exceptions enabled due to included custom ops "
+                 "using 3rd party libraries that require exceptions.")
+
+  if (NOT OCOS_ENABLE_CPP_EXCEPTIONS)
+    message(WARNING "Enabling C++ exception support as custom ops included in the build require.")
+    set(OCOS_ENABLE_CPP_EXCEPTIONS ON)
+  endif()
+
+  # undo the flags that ORT has set to disable exceptions.
+  # see https://github.com/microsoft/onnxruntime/blob/b1abb8c656c597bf221bd85682ae3d9e350d9aba/cmake/adjust_global_compile_flags.cmake#L160-L169
+  if(MSVC)
+    string(REPLACE "/EHs-c-" "/EHsc" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+  else()
+    string(REPLACE "-fno-exceptions" "-fexceptions" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+    string(REPLACE "-fno-unwind-tables" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+    string(REPLACE "-fno-asynchronous-unwind-tables" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+  endif()
+
+  # the ort-ext code has to provide a barrier between the exception enabled custom op code and ORT.
+  add_compile_definitions(OCOS_PREVENT_EXCEPTION_PROPAGATION)
+endif()
+
 if(NOT OCOS_ENABLE_CPP_EXCEPTIONS)
```

Comment:
Do we still need to report the error here? since C++ exceptions can be partially disabled for some operators now.

Reply:
Possibly not, but not sure about one aspects of the setup though.

We're building some operators with and some without exceptions.
We have ORTX_CXX_API_THROW to provide a consistent way to throw, but that is dependent on OCOS_NO_EXCEPTIONS which is not operator specific. 
Operators have no way to report errors in construction or compute with the current ORT custom op API having no ability to return a status, so AFAIK they MUST throw (or log+abort if execeptions disabled) for any error.

What is the expected operator implementation for something that compiles without exceptions? The string operators currently use ORTX_CXX_API_THROW but I think are in the operators we build with exceptions disabled so not sure how that is working.

Reply:
Is the current setup if exceptions are disabled as follows:
- set OCOS_NO_EXCEPTIONS 
  - which causes all calls to ORTX_CXX_API_THROW to log+abort
- compile the noexcep_operators operators with adjusted flags to disable exceptions
  - this code can use ORTX_CXX_API_THROW which will result in log+abort
- compile other operations with exceptions enabled
  -  3rd party code can call `throw`

For the operators that throw in a build with exception disabled what is the expected behavior? Propagate exception up (no call to `abort`), or something like OCOS_API_IMPL_END should catch the exception and log+abort so all operators have consistent behavior?

Reply:
My initial thinking is simple:
If OCOS_ENABLE_CPPEXCEPTION is OFF, then all code in ort-extensions will NOT throw any exception since ORTX-THROW will abort immediately if something unexpected is happening.
But try-catch macro in API function is effective and it actually just only catch the exception from 3rd party libraries.

There is one thing missing here is that BaseKernel should have an exception-free superclass, which will be the base class for non-exception operators. So if only non-exception operators are compiled, there actually no try-catch code in whole compiling source code, I guess the binary size is still minimal in that case.

Reply:
Updated to also log+abort in OCOS_API_IMPL_END if exceptions are disabled.

Not sure it's worth the complexity of a BaseKernel layer without testing to see if the binary size saving justifies it and there a use case that requires it.